components/net/decoder.rs

```diff
@@ -0,0 +1,508 @@
+The decompressed chunks aren't guaranteed to align to the compressed ones.
+
+If the response is plaintext then no additional work is carried out.
+Chunks are just passed along.
+
+If the response is gzip, then the chunks are decompressed into a buffer.
+Slices of that buffer are emitted as new chunks.
+
+This module consists of a few main types:
+
+- `ReadableChunks` is a `Read`-like wrapper around a stream
+- `Decoder` is a layer over `ReadableChunks` that applies the right decompression
+
+The following types directly support the gzip compression case:
+
+- `Pending` is a non-blocking constructor for a `Decoder` in case the body needs to be checked for EOF
+- `Peeked` is a buffer that keeps a few bytes available so `libflate`s `read_exact` calls won't fail
+*/
+
+use brotli::Decompressor;
+use bytes::{Buf, BufMut, BytesMut};
+use crate::connector::BUF_SIZE;
+use flate2::read::DeflateDecoder;
+use futures::{Async, Future, Poll, Stream};
+use hyper::header::{HeaderValue, CONTENT_ENCODING, TRANSFER_ENCODING};
+use hyper::{self, Body, Chunk, Response};
+use libflate::non_blocking::gzip;
+use servo_url::ServoUrl;
+use std::cmp;
+use std::fmt;
+use std::io::{self, Read};
+use std::mem;
+
+pub enum Error {
+    Io(io::Error),
+    Hyper(hyper::error::Error),
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error::Io(err)
+    }
+}
+
+impl From<hyper::error::Error> for Error {
+    fn from(err: hyper::error::Error) -> Error {
+        Error::Hyper(err)
+    }
+}
+
+const INIT_BUFFER_SIZE: usize = 8192;
+
+/// A response decompressor over a non-blocking stream of chunks.
+///
+/// The inner decoder may be constructed asynchronously.
+pub struct Decoder {
+    url: ServoUrl,
+    inner: Inner,
+}
+
+#[derive(PartialEq)]
+enum DecoderType {
+    Gzip,
+    Brotli,
+    Deflate,
+}
+
+enum Inner {
+    /// A `PlainText` decoder just returns the response content as is.
+    PlainText(Body),
+    /// A `Gzip` decoder will uncompress the gzipped response content before returning it.
+    Gzip(Gzip),
+    /// A `Delfate` decoder will uncompress the inflated response content before returning it.
+    Deflate(Deflate),
+    /// A `Brotli` decoder will uncompress the brotli-encoded response content before returning it.
+    Brotli(Brotli),
+    /// A decoder that doesn't have a value yet.
+    Pending(Pending),
+}
+
+/// A future attempt to poll the response body for EOF so we know whether to use gzip or not.
+struct Pending {
+    body: ReadableChunks<Body>,
+    type_: DecoderType,
+}
+
+/// A gzip decoder that reads from a `libflate::gzip::Decoder` into a `BytesMut` and emits the results
+/// as a `Chunk`.
+struct Gzip {
+    inner: Box<gzip::Decoder<Peeked<ReadableChunks<Body>>>>,
+    buf: BytesMut,
+    url: ServoUrl,
+}
+
+impl fmt::Debug for Decoder {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("Decoder").finish()
```

Comment:
Not sure what this is exactly doing but I don't think it gives the type of decode and it might  be useful to know it ? i.e plain, gzip...